docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,428 @@
+validation failing on the receiving chain.
+
+2. The existence of special characters and uppercase letters on the denomination:
+
+In the SDK every time a `Coin` is initialized through the constructor function `NewCoin`, a validation
+of a coin's denom is performed according to a
+[Regex](https://github.com/cosmos/cosmos-sdk/blob/a940214a4923a3bf9a9161cd14bd3072299cd0c9/types/coin.go#L583),
+where only lowercase alphanumeric characters are accepted. While this is desirable for native denominations
+to keep a clean UX, it presents a challenge for IBC as ports and channels might be randomly
+generated with special and uppercase characters as per the [ICS 024 - Host
+Requirements](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements#paths-identifiers-separators)
+specification.
+
+## Decision
+
+Instead of adding the identifiers on the coin denomination directly, the proposed solution hashes
+the denomination prefix in order to get a consistent length for all the cross-chain fungible tokens.
+The new format will be the following:
+
+```golang
+ibcDenom = "ibc/" + hash(trace + "/" + base denom)
+```
+
+The hash function will be a SHA256 hash of the fields of the `DenomTrace`:
+
+```protobuf
+// DenomTrace contains the base denomination for ICS20 fungible tokens and the source tracing
+// information
+message DenomTrace {
+  // chain of port/channel identifiers used for tracing the source of the fungible token
+  string trace = 1;
+  // base denomination of the relayed fungible token
+  string base_denom = 2;
+}
+```
+
+The `IBCDenom` function constructs the `Coin` denomination used when creating the ICS20 fungible token packet data:
+
+```golang
+// Hash returns the hex bytes of the SHA256 hash of the DenomTrace fields.
+func (dt DenomTrace) Hash() tmbytes.HexBytes {
+  return tmhash.Sum(dt.Trace + "/" + dt.BaseDenom)
+}
+
+// IBCDenom a coin denomination for an ICS20 fungible token in the format 'ibc/{hash(trace + baseDenom)}'. If the trace is empty, it will return the base denomination.
+func (dt DenomTrace) IBCDenom() string {
+  if dt.Trace != "" {
+    return fmt.Sprintf("ibc/%s", dt.Hash())
+  }
+  return dt.BaseDenom
+}
+```
+
+In order to trim the denomination trace prefix when sending/receiving fungible tokens, the `RemovePrefix` function is provided.
+
+> NOTE: the prefix addition must be done on the client side.
+
+```golang
+// RemovePrefix trims the first portID/channelID pair from the trace info. If the trace is already empty it will perform a no-op. If the trace is incorrectly constructed or doesn't have separators it will return an error.
+func (dt *DenomTrace) RemovePrefix() error {
+  if dt.Trace == "" {
+    return nil
+  }
+
+  traceSplit := strings.SplitN(dt.Trace, "/", 3)
+
+  var err error
+  switch {
+  case len(traceSplit) == 1 && traceSplit[0] == dt.Trace:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must contain '/' separators", dt.Trace)
+  case len(traceSplit) == 1:
```

Comment:
I don't think this case occurs because if there exists a slash in `dt.Trace` there should be a `traceSplit` of length 2 otherwise the first case is hit

Reply:
we can probably remove it since the denom trace is validated on the `msg.ValidateBasic`

Reply:
I think we can still have the ` case len(traceSplit) == 1 && traceSplit[0] == dt.Trace:`, I just meant that the `case len(traceSplit) == 1:` will never be hit because the former will always bit hit if len == 1